Remove most of C++ build warnings

MartinBriza opened the pull request:
### 📒 Description
With this PR, we get rid of most of compiler warnings with msvc, clang and g++. For a more comprehensive list of what has been changed, you can check the individual commits, I did a separate one for each type of change.

The most interesting ones are:
 * Passing `std::string` into `copy_string` istead of using  `c_str ` (avoids one unnecessary copy)
 * Deprecated Qt method changes (without change in behavior)
 * Passing const references instead of values for vectors and strings (because of possibly improved performance)
 * There was actually a small bug with autotracker rules when they were not stored as lowercase (and an unused variable was left behind) 

### 🕶️ Types of changes
- **Bug fix** (non-breaking change which fixes an issue) 

### 👫 Relationships
Closes #4538 

### 🔎 Review hints
Everything should work the same as before, except:
 * No library C++ build warnings should occur on macOS
 * Only QNetworkConfiguration related ones on Linux (#4539)  
 * There should be less of them on Windows (I don't really understand why some of them happen in there)

atsvirchkova reviewed (changes requested):
There have become a lot less warning than before, thank you for this job. But according to review hints and build logs there is something left, which should not be left:
```
/home/runner/work/toggldesktop/toggldesktop/src/color_convert.cc:157:23: warning: ‘b’ may be used uninitialized in this function [-Wmaybe-uninitialized]
  157 |     return { r , g, b };
      |                       ^
/home/runner/work/toggldesktop/toggldesktop/src/color_convert.cc:157:23: warning: ‘g’ may be used uninitialized in this function [-Wmaybe-uninitialized]
/home/runner/work/toggldesktop/toggldesktop/src/color_convert.cc:157:23: warning: ‘r’ may be used uninitialized in this function [-Wmaybe-uninitialized]
```
from linux build,
get_focused_window_mac.cc contains some warnings, that something is deprecated (macOS build)
```
/Users/runner/work/toggldesktop/toggldesktop/src/toggl_api.h:352:40: warning: this function declaration is not a prototype [-Wstrict-prototypes]
    typedef void (*TogglContinueSignIn)(
```
also macOS build.

MartinBriza commented:
Thanks for the review, Anna, I addressed the remark about the colors being initialized, however I'm convinced it was not necessary because the `switch` below has covered all 6 possible cases. To be 100% sure, I added an explicit initialization to those values, though.

As for this:
```
/Users/runner/work/toggldesktop/toggldesktop/src/toggl_api.h:352:40: warning: this function declaration is not a prototype [-Wstrict-prototypes]
    typedef void (*TogglContinueSignIn)(
```
The prototype is now: 
```c++
    typedef void (*TogglContinueSignIn)();
```
And a proper solution to this is changing it to:
```c++
    typedef void (*TogglContinueSignIn)(void);
```
There's no real difference between those two but the compiler likes one of them more for some reason. I'm not changing it simply because I suspect it would break `make csapi` and I don't have Windows ready to fix that at this moment. I'm opening an issue for this as a possible future improvement: #4561

atsvirchkova reviewed (approved):
Good job 👍  Merging it now.
